### ISSUE TITLE
Age Verification

### ISSUE THREAD

**Issue by DBailey635** [mastodon/mastodon#24440]
### Pitch

Mastodon may need a way to verify the ages of users where required by law. There should be some option on account creation that lets instance owners redirect users to a third-party age-verification provider and accept their tokens in order to proceed. Having such a facility would also allow a method of screening NSFW content marked as such (see #9468 ).

Additionally, instance owners should be able to restrict access from certain regions if they are unable to implement such controls (due to the associated costs).

### Motivation

Several countries and US states have enacted age verification requirements for social media sites. This means teens would not be permitted to access social media sites without either proving they are over specified age or providing parental consent.

**Comment by DBailey635** [comment]
Example: Utah: https://www.engadget.com/utah-passes-laws-requiring-parental-permission-for-teens-to-use-social-media-223302739.html

**Comment by DBailey635** [comment]
@GBT7 This is why the mechanism I'm suggesting is to use an external provider for age verification. Instance admins won't have anywhere near the resources to do it themselves, so they should let someone else do it on their behalf.

The annoying thing about the legislation being passed, is that often it doesn't differentiate between social media sites with tens or millions of users. Just adding questions to the site won't be enough to satisfy the law... it often requires proof in the form of State-approved ID.

On the subject of requiring ID to access porn or NSFW sites, it's coming soon in Europe, and will probably be required in some US states before long. See: https://www.wired.co.uk/article/porn-age-checks-id-laws 

Whatever happens, Mastodon is going to have to run this past its lawyers to see what they recommend, in order to help protect all instance admins. 😔

**Comment by ErikLentz** [comment]
> The annoying thing about the legislation being passed, is that often it doesn't differentiate between social media sites with tens or millions of users.

That's often a part of any regulatory effort: stifle competition via regulations that only larger entities can accommodate.

**Comment by JustDalek** [comment]
+1 for age checkboxes as suggested by GBT7

I already have a rule in place for my server, would be cool if the rule itself had a checkbox that was required next to it

**Comment by lukefromdc** [comment]
I for one will refuse to use any complying server. A perfectly legal alternative is to block every IP address in the offending states or countries, then welcome Tor connections. That's what Pornhub does about age verification laws: block and welcome Tor and VPN access.

So long as there are noncomplying (or blocking overt access) servers, that where many of us will go. I am so opposed to these laws I would sooner disconnect from the Internet entirely than comply. Any Mastodon instance that has "no business presence" in an offending state or country should be able to ignore this.If not, see the Pornhub approach above.  A hole in these laws is that thankfully they are not written with Mastodon in mind. A person migrating to a noncomplying server can thus retain any followers from a complying server, probably with no risk to the complying server.

As a rule, I will never permit any site to verify my age, I do not carry ID in person and will not retain any account on any server asking for ID. This is an extension of my longstanding refusal to have accounts on sites that demand SMS verification. I will always block all 3ed party age verification servers from connecting to any device I own, just as I now block Google, Facebook, and all of the ad networks.

In the long run, the Internet may divide into two opposite portions: legal, censored, and monitored sites in the China model, and the darknet with nothing in between. If this happens I will be found only on the darknet. 

If Mastodon goes down over this, distributed hosted Diaspora would be the obvious replacement, just as torrents replaced the orginal Napster filesharing service. Peer to peer, distributed hosted models have no central server, they are a fleet with no vulnerable flagship to attack. They are very difficult for censors to take down

**Comment by MirceaKitsune** [comment]
Age segregation... I mean discrimination... sorry I mean verification... has no place and will never work in an open environment like the Fediverse, and in my book even considering it is the equivalent of blasphemy to both this project and open source in general. Apart from this being a plain wrong thought-police practice done over ridiculous superstitions that sum up to "pixels on a screen are killing children", it's not technically feasible and any attempts to have such a thing would literally end everything Mastodon tried to be, the team may as well shut down the project at that stage. Indeed we live at a day and age where any crazy person can take over a government and try to force insane laws, many of which no one will realistically accept or even be able to comply with: Let us keep hoping the era of madness burns itself out soon, and in the meantime let's improve and defend what's left of the actual internet keeping it free and open for everyone.

**Comment by lukefromdc** [comment]
My hope is that complying servers would shut down from migration away from them while noncomplying servers hosted where these goons csnnot easily reach keep going.

As governments come for the Internet, we may need to think about ways to support Tor-only instances and federate them with each other but never expose their ownership.

**Comment by hikayamasan353** [comment]
I think that we should actually implement parental consent mechanism as well.

While I understand your desire to comply with COPPA, the big and unintended long term consequences of COPPA is that almost every social media platform just banishes kids under 13, which obviously has resulted in kids having to either wait until 13 and lie about their age, with or without their parents involved.

If we focus on just banishing kids, we are depriving them of opportunities that can be very beneficial for them. Since kids under 13 have none, they're essentially isolated and voiceless. It's a clear case when focus on children's online privacy and safety has compromised basic children's rights.

In fact, social media services banish kids not to protect them. They do because they avoid legal burdens of obtaining parental consent. This doesn't serve kids, it serves social media services. 

Here's a real life analogy. Recently, playgrounds have become too safe. Classic playground equipment such as swings, seesaws, monkey bars, merry go rounds, high slides etc are banned. Instead, slides are too low, the surface is rubber padded everywhere. All this is driven by lawsuits over playground incidents, and fear of them. Research has shown that a risk-free environment is detrimental to children's development, and a certain level of risk is actually beneficial for them. In fact, as a response to the trend of playgrounds getting too safe, "adventure playgrounds" have been introduced, where kids have to use real tools and junk to build playground equipment.

**Comment by lukefromdc** [comment]
I can guarantee you some servers (notably hard core activists server)
will refuse to implement any sort of age verification or parental consent
measures. I myself will always be on such servers and will never verify
age or identity for any site.

For some servers serving communities disfavored by governments, it can
be an important security measure NOT to retain any records that could
be useful for an adversary attempting to prove ownership of an account.

This could lead to a "fedisplit" where the issue causes a bifurcation of the
community, with anonymity servers federating mostly with oneanother
and being blocked by identity verifying servers. I know which side of that split
I will always be on. I would even be OK with moving my work to a tor-only
server that federated only with other torified servers if necessary to evade
legislation demanding age and identity verification.

That said, if someone wanted to set up a subnetwork specifically for kids,
I would imagine it would not take a lot of modification of mastodon for
that fork to work.

**Comment by DBailey635** [comment]
Now the UK may be getting involved:  "All websites on which pornographic material can be found, **including social media platforms**, must introduce "robust" age-checking techniques".

[https://www.bbc.co.uk/news/articles/cwye3qw7gv7o](https://www.bbc.co.uk/news/articles/cwye3qw7gv7o)

**Comment by MirceaKitsune** [comment]
Open systems like the fediverse can't and shouldn't try to appease third-world nations with authoritarian regimes that don't understand how the internet works. Russia has a system for blocking websites the government doesn't like at ISP level, not to mention China which walled itself off the global internet entirely: That doesn't mean every service on the web will compromise itself for their hubris, we don't change our design to give Putin or Xi special privileges so neither do we for Tory / Labor. Ignore those places as there's nothing we can do for them unfortunately, other than hoping they'll see better days again and be ruled by sane people eventually.

**Comment by lukefromdc** [comment]
> Now the UK may be getting involved: "All websites on which pornographic material can be found, **including social media platforms**, must introduce "robust" age-checking techniques".
> 
> https://www.bbc.co.uk/news/articles/cwye3qw7gv7o
> 
> .... and if they don't, they may be blocked at the ISP level.

Let them block! Torbrowser will burn through those blocks like a hot knife through a stick of butter. So will many VPN's.

For those who themselves are subject to these sort of laws based on where they live and are placed at risk by running noncompliant servers, I favor Pornhub's model: pre-emptively block all IP addresses originating from the offending jurisdictions, them welcome Tor and VPN connections. Pornhub works just fine in Florida without ID and over Tor. Tested it myself, it works just fine.

Pornhub's response to these laws is actually going to help build VPN usage, build the Tor network, and induce millions to use Tor for the first time. This is a beneficial effect that makes all other censorship more difficult. Enough of this and the whole Internet "goes dark" to governmental and corporate spies, with nobody able to trace anyone's online activity. Porn and social media will be the "killer apps" for Tor as Visicalc was the "killer app" for the Apple II and putting microcomputers on offices so many years ago.

As for me, I know where I stand: I will not be part of any community that implements age or identity verification, no exceptions.

**Comment by wjeffreybrown** [comment]
All we want is a simple checkbox that says.  **I confirm that I am "N" Years of age.** during the registration. 

That would satisfy most of these requirements.    Years later, it's still absent.  As is a separate Terms of Service page.

**Comment by MirceaKitsune** [comment]
> All we want is a simple checkbox that says. **I confirm that I am "N" Years of age.** during the registration.
> 
> That would satisfy most of these requirements. Years later, it's still absent. As is a separate Terms of Service page.

The registration page can already display a customizable set of rules right? For people and places with such concerns, they can just list that stuff there if they really feel a need to.

I don't see any reason to also have a checkbox; If some really want an extra step, you could add one of those generic "I have read and accept the rules" boxes. But I don't see what use that is and just find them annoying: You click the continue button anyway, it's just interface clutter making you click twice instead of once for the same thing. I think most of us are happy with everything being simple and straightforward: Drama over weird social issues is already embedded in pretty much every instance, it doesn't need to also be in the code by default.

As for things that are absent years later, we've all been asking for a way to import our archives: Almost everyone wants it yet there's still no sign from the devs adding support years later. Remember that if an admin really wants a missing feature, Mastodon can be forked with plenty of forks already out there and used by some servers.

**Comment by hikayamasan353** [comment]
You didn't read my point of view. Why thumbs down???

> All we want is a simple checkbox that says. **I confirm that I am "N" Years of age.** during the registration.
> 
> That would satisfy most of these requirements. Years later, it's still absent. As is a separate Terms of Service page.

But this will never satisfy the reality that kids under 13 are already online, and parents are actually helping them to create accounts and also manage them too.

But you will be like: "But kids will see mature content online", "They will meet a pedophile who will kidnap them over the Pacific and murder them overseas"...

I'm sorry but these points are null.

Even if you keep kids under 13 off all social media and Internet as a whole "for their own good", this doesn't mean they're going to be safe or protected.

Besides the Internet, there are many other places children can see inappropriate content or witness inappropriate behaviour. In addition, so harder we shield kids from inappropriate content, so more curious they will be about it. This is exactly why Elsagate videos have gone viral. We don't avoid talking about sex with kids or tell kids that storks brought them or that tampons/pads are "something else" until puberty. But being open and honest with children about sexuality in an age appropriate way is very different from showing kids porn or other adult material.

The fear of a pedophile contacting a child online and then meeting them up offline to kidnap and murder is nothing but overblown "stranger danger" narrative. It has been statistically proven that most of the harm to the children comes from the people who are ALREADY KNOWN to the child, or who already know the child - implying some degree of familiarity. Besides, the most common inappropriate behaviour online that kids can come across, is blatant cyberbullying, petty grooming or general toxicity, rather than someone deliberately asking where do you live or wanna meet up immediately - it's a straightforward red flag, that's why grooming is much more common and treacherous. However, there's no need for paranoia - grooming involves deception and betrayal of trust.

In addition, COPPA was NOT designed to address these problems. It is a privacy law, and the reason why it was drafted back in the 90s was a simple privacy leak.

If you don't want kids to use social media, then I'm sorry but they already use landline phones (if we either ignore cell phones or assume that kids have no cell phones) and snail mail - by the way, sending letters to Santa collects much more sensitive and identifiable data than registering online.

If you think that kids don't need social media, then what about all those kids who are frequently rejected and bullied by peers, who live in areas with very few other kids to hang out with offline, etc? Do you really believe that kids should just deal with it and wait until 13 before even trying to say hi to someone who is not their relative or family member, online or offline?

**Comment by lukefromdc** [comment]
Some of the more underground servers might choose to refuse to implement even an
"age of majority flag," I for one would not want that on my personal profile as I perfer
to provide cover for those who do have to lie about their age.

Fair disclosure: I oppose the OBJECTIVE of all these new legislative proposals and
will always favor instances that rely on making it hard to physically find or reach the
admins for defense rather than attempting to appease ever nastier politicians.

**Comment by hikayamasan353** [comment]
@GBT7 
> It's just to shift the liability away from instance admins, that's all it serves to do.

This is why kids have literally none at all, or few if any, online spaces they can hang out, make friends etc on. **I already said this [before.](https://github.com/mastodon/mastodon/issues/24440#issuecomment-2294988331)**  **The very intention to shift the liability** only shows that no online server operator (including Mastodon instance admin) even thinks about children's online presence and their personal information. To them, kids under 13 simply don't belong on the Internet at all, and that's why they see no reasons to even think about trying to let them in, which is only, under COPPA/GDPR, possible by implementing parental consent.

What's worse, parental consent can be implemented using standard e-commerce techniques (according to 312.5b1) - they almost always satisfy the requirement of verifying the parents' identity:

> An [operator](https://www.law.cornell.edu/definitions/index.php?width=840&height=800&iframe=true&def_id=fc96bc55e64efe6c69a0d240fa01982c&term_occur=999&term_src=Title:16:Chapter:I:Subchapter:C:Part:312:312.5) must make reasonable efforts to obtain verifiable parental consent, taking into consideration available technology. Any method to obtain verifiable parental consent must be reasonably calculated, in light of available technology, to ensure that the [person](https://www.law.cornell.edu/definitions/index.php?width=840&height=800&iframe=true&def_id=afab0b6cea66321581211516f7d7157e&term_occur=999&term_src=Title:16:Chapter:I:Subchapter:C:Part:312:312.5) providing consent is the [child](https://www.law.cornell.edu/definitions/index.php?width=840&height=800&iframe=true&def_id=81dce68d6fa4724c7029bf874fadc422&term_occur=999&term_src=Title:16:Chapter:I:Subchapter:C:Part:312:312.5)'s [parent](https://www.law.cornell.edu/definitions/index.php?width=840&height=800&iframe=true&def_id=009c11fb9d9acba4202d58a7fec13fb0&term_occur=999&term_src=Title:16:Chapter:I:Subchapter:C:Part:312:312.5). **(16 CFR 312.5b1)**

The examples cited in the next part (312.5b2), at first glance can seem intimidating and expensive (toll free phone number staffed by trained personnel, postal mail, fax, government ID scan etc). Hearing "toll free phone number staffed by trained personnel" can indeed drive their fear and shock, because: 

- Toll free phone calls are paid by recipient (unlike the generic phone calls where the caller pays), and the subscriber fee for a toll free phone number is higher than for a generic phone number. 
- Personnel training is not only expensive but there is few if any curricula on training phone operators specialized in obtaining parental consent. Generic call center operators might be unfamiliar with this.

The requirement of a toll free phone number staffed by trained personnel, as an approved method of parental consent, is a very high bar. Luckily, FTC doesn't require just that - all they require is simply so that the method of obtaining parental consent is just enough to verify the parents' identity.

The reality is that many of the potential methods that are sufficient enough to verify parents' identity and therefore satisfy the 312.5b1 requirement of COPPA , can be used and are already widely used in e-commerce, even if they are not listed in the 312.5b2, which is non-exhaustive and non-exclusive anyways.

Credit cards, listed as an example, in the current age of freemium SaaS subscriptions, can freely be used to verify identity, especially when they are already connected to the identity of an adult. Credit cards and printable/digital forms are very dirt cheap to implement, unlike a toll free phone number staffed by trained personnel.

Yet, to almost every online service operator, as well as even some FTC officials, parental consent means a Randian challenge that only can be passed by breaking fundamental laws of physics. Especially given the fact that COPPA fines are so high they can put a small business bankrupt. **Which is why online service operators (including Mastodon instance admins) choose to simply evade liability by keeping kids under 13 off their services.**

**And this evasion of liability hurts kids and parents.** It erases their online presence. No wonder why I call COPPA "Children's Online Presence Prevention Act".
This was, is and will always be my point.

**Comment by lukefromdc** [comment]
This in turn is another reason why its so important to kill age verification: so service providers can have
plausible deniablity. If that doesn't work instances in the offending jurisdictions become unusable for both
kids and those concerned with privacy, those that are hosted outside the offending jurisdictions by people
who do not have passports/do not travel internationally are immune. 

Also note that since a Mastodon instance is normally not monetized it is not a business and has zero revenue.
This repeats the scenario that MPEG-LA faced with ffmpeg and users who published videos using H264 in 
defiance of US software patents. Distros that were not businesses and 3ed party repos hosted patent-busting
builds of ffmpeg and users ignored the patents until MPEG-LA issued a blanket license for non-monetized video
posted by end users.

I am very much from the end of the spectrum where oppressive laws are defied and treated with the contempt
they deserve, and where it is accepted that defense can become an issue. Real worst case, .onion-only Mastodon
instances with no verification of anything can be set up. If age verification was hardcoded into open-source mastodon,
anyone could revert the offending commits and build without them.  I think the internet as a whole is going to split
between a "main street" regulated and surveilled "brightnet" and a greatly expanded darknet.

**Comment by dsntechadmin** [comment]
Hi

I am an Australian Content Creator and Broadcaster, with a PeerTube server, CSPTube.

In Australia, it now needs to be ID from December 10th 2025, or I need to close the server,. and I DO NOT WANT THAT!

So we need at least a quick ID checker, in association with a AI or S3 bucket, to quickly check the IDs, and then IMMEDIATELY delete that information, for the subscribers privacy.  What is happening in the world is BS, and now I am starting to compose a petition, to delay this new Age Verification laws for 6 months.

We also have a problem with Collective Shout.. because.. they are douchebags.

**Comment by lukefromdc** [comment]
My advice would be to move the server out of range of your country's law enforcement and if 
necessary conceal the ownership. Be advised many in the Fediverse may refuse to continue
federating with any Mastodon, Peertube or other server that has age verification in use.

**Comment by scytaleqc** [comment]
Implementing age verification seems like a lot of work to comply with demands that ultimately, go against the very principles the fediverse is based upon.

Age verification and KYC laws fundamentally centralize power in a few hands. This is the opposite of the stated goals of the fediverse.

Really not sure how this situation should be adressed but I will not track my users or be part of a network that does.

**Comment by ronanhigg** [comment]
Open letter to Mastadon age-verification working group, from Veritage:  
(This letter already sent to Felix Hlatky) 

...in relation to the [vision, mission and strategy memo yesterday](https://blog.joinmastodon.org/2026/02/connecting-the-world-through-thriving-online-communities/) announcing that Mastadon is “forming an interdisciplinary age-verification working group to ... navigate the complex nature of ... international laws.“...

**Veritage** would like to help you navigate.

We would like to integrate our age assurance solutions into the Mastadon server code base as an optional feature, so that Mastadon server owners can, with just the flip of a switch, make their server compliant with age verification regulations across the world.

We could have this issue solved for the Mastadon community quickly. 



In contrast to age verification providers like Yoti, Persona and k-ID, ...
**Veritage** has focused on building a suite of _privacy-forward_ solutions that ensure social network users do not feel “betrayed” when asked to prove they are above the legal age required in their jurisdiction. 

We don’t believe it is safe to upload passports to servers or stream face embeddings to the likes of Palantir & Co.  And we don’t believe Mastadon servers should require this of its users, when there are superior technologies available to comply with regulations.


**Veritage** offers a comprehensive age verification solution:

- Geography-based regulatory compliance
- VPN usage detection
- Privacy-forward “zero knowledge proof” (ZKP) age verification
- Frictionless and secure client-side AI-selfie face age estimation (FAE) (optional)
- PassKey verification storage (optional)
- Support for Digital Credentials such as EU Digital Identity wallet, Google Wallet and Apple Wallet (optional)


We are happy to invest the time and development resources to commit Veritage to the Mastadon server code base, similar to how CAPTCHA has been made available as an option in Server Settings / Registrations.

Instead of a the current self-declaration of date of birth on registration, users would not need to provide a date of birth at all.   The zero knowledge proof would satisfy the legal requirements that the registrant is of age of majority.

This I believe is the path forward.

...

Regards, 
Ronan Higgins
Co-founder 
ronan [a t] veritage.io

**Comment by colinstu12** [comment]
Masto being sold off to the highest bidder. Fucking sick.

**Comment by lukefromdc** [comment]
Any age-verified account on any server risks proving ownership of every post on every platform made
by that user by a standard good enough for divorce court, good enough for ICE, the works. I understand
some servers will do this anyway, some will shut down, and some are out of reach of any nation's 
age verification laws or enforcement therof.

These choices serve people with different needs. Those of us with high security needs such as myself 
however cannot use any site or service regardless of its nature that verifies age or any other aspect
of identity..  Thus I will only ever be found on noncompliant servers, I expect at least two years before 
complying servers face effective harassment  for federating with non age verifying servers. There is literally
nothing anywhere online I am sufficently interested in to show ID to get it. I apply the same standard
if a chain store asks for ID to buy Super Glue or paint: "get rid of it," and I have never had any store
give me any problem about having them reshelve the offending item and continuing without ID.

If all of Mastodon was forced to do this by an upstream software change with no admin option to use it
or not, I would close my account. Note that I am not on mastodon.social (the one currently known to
need this) as that server is unsuited to my needs.

If few servers are able to avoid age verification but federation is not interfered with, I will be advising users
to set up their own single-user instances, if necessary self-hosting them to avoid web hosts that want
identity verification. A single user instance by definition does not host (though it can display) 3ed party
content. Also note that small servers in partially-repressive jurisdictions may not have enough users
to trigger "large platform" age verification mandates in those areas. 

Note that as of now, at least some Mastodon instances (including the one I use) have .onion access.
It would be very useful to be able to create .onion-only instances and connect them to other instances
also over tor using .onion to avoid the exit node bottleneck. The darknet is the wave of the future with
the rapid growth of online censorship and surveillance. A network of .onion-only instances that do not
use any clearnet connections at all could defy the censors almost with impunity. Finding them to take them
down would be difficult and resource-intensive. I think it took the FBI over a year to shut down Silk Road-
JUST Silk Road.

**Comment by ronanhigg** [comment]
Hi @lukefromdc , thanks for your comment on this issue.   Would you mind elaborating on your first point.  "risks proving ownership of every post on every platform made by that user by a standard good enough for divorce court, good enough for ICE, the works."   

I don't understand how that would happen, just because an account is ageOver16=true.

On your second point about not wanting to show ID online or offline, I totally get that.  It's kind of insane that people have to present a driver license to a bouncer, bar tender, shop assistant to buy age restricted product, when that card has your home address, name, and date of birth on it. Serial murderers and stalkers love it.

That's the point of zero knowledge proof (ZKP) age verification with self-sovereign identity (SSI) wallets.  No personally identifiable information is presented.  A zero knowledge proof simply says "I am over 16".  No name, no date of birth, no photograph, no address.   To use your analogy of buying Super Glue or paint, if they are age restricted products, and the shop assistant asks you to prove you are over the required age, you don't have to present an identity document to the shop assistant.   You would simply say "I am over 16" and he would take that statement at face value.   Why would he believe you?  Because cryptography.   I respect that that sounds facetious, but zero knowledge proof cryptography is kind of like a form of magic.   

This means that a server, or a social network, or an online Super Glue store, could ask a user "are you over 16?" and the user would send a zero knowledge proof that simply says "I am over 16" -- with no identifying information -- and the server would accept that as truth.

Because of the dearth of information passed over, it is not possible to link that user to anything.  Unlinkability is at the core of the system.   

I know it sounds a bit too good to be true.   When I learned about this area of mathematics last year, I was amazed.   I've become so used to an Internet where data is promiscuous and hoovered up by corporations and governments.   So when I saw how a zero knowledge proof works, a light switched on.   This has the potential to address so many privacy issues, especially for people who have high security needs.   

I recommend you take a read through some of the literature about zero knowledge proofs and self-sovereign identity.   It is the path forward to feeling safe buying Super Glue and paint, online and offline.

**Comment by lukefromdc** [comment]
In today's world, to transmit an ID online to anywhere risks spyware in machimes on either end of the transmission copying it. We have confirmation that ICE is using Paragon's Graphite spyware in the US to target migrants. Graphite is probably easily able to do this, right on the sender's phone prior to encryption for transit.

Also anyone using an age verified site is probably going to be on at least one ad supported site. These sites build dossiers on all users that they can, meaning everyone that doesn't eirher boycott them or deploy very robust af and tracker blocking. One site in this dossier verified to an identity (including by ICE with intercepted data) also verifies all other sites the ad nrtwork data can confirm as being from the same user. Even a birthday can segment a data set into 365 much smaller groups.

Not only do I not transmit ID over any network, I also block ads on all sites and block both Google and Meta at the toplevel.

**Comment by trwnh** [comment]
> zero knowledge proof (ZKP) age verification

this assumes the problem with age verification is info leakage and not the verification itself.

"the path forward to feeling safe buying Super Glue and paint, online and offline" is to not require any verification whatsoever of any information beyond perhaps payment details. when it comes to communication and speech, there shouldn't be a forbidden class of knowledge. the existence of such a class cannot be trusted in a world where legislators are trying to classify sex education, abortion resources, and even queer people's existence as 18+. the only safe age verification is no verification. implementing age verification is not a safety measure; it is the abdication of any responsibility to keep people safe.

this is not legal advice, but if given a choice between complying with laws like this vs not complying, i would advocate everyone to not comply as much as possible. these laws need to be ridiculed and repealed, not respected. civil disobedience and resistance is the correct response.

**Comment by lukefromdc** [comment]
I actually advocate defiance of unjust laws and will favor servers that defy this no matter how implemented. I underarand this will probably be a small fracrion of total servers bur it DOES include most of the activist-run servers presumably.

I will be right up front: I oopose the OBJECTIVES of age verification, but for some folks the security implications of using ID are even bigger. The objectives are also harmful, notably to LGBTQ yourh who Christian nationalists are trying to isolate. Beat way to protect them is to go hard no-compromise/no quarter on issues rhat could give governments the power to cut their comms.

No compliant platform will be part of my online life, just as I buy Super Glue only in person with cash and wirhout ID. I would use different glues before I would change that. I'm not saying non-activist servers cannot comply to avoid the defense requirements of running an "underground" server, rather that I will not be found on any complaint server. Even if noncompliant servers cannot be federated by complaint ones and the Fediverse splits in two, I will be on the resisting/defying side of this.

If ALL clearnet platforms implement this, I will abandon all of them and publish my video work only on the darket and encrypted messaging (which can be done peer to peer if necessary). I exited the corporate platforms 15 years ago and am prepared to do the same wirh all of clearnet if I have to.

**Comment by lukefromdc** [comment]
Also note this study:
csmapnyu.org/research/academic-research/do-age-verification-bills-change-search-behavior-a-pre-registered-synthetic-control-multiverse

Which shows broad user bypassing of age verified sites and search behavior shifting to noncompliant sites. Plenty of those, often hosted in countries out of reach of the offending governments. Legal starus is identical ro that of Google refusing to implement China's censorship demands, welcoming Tor and VPN traffic from China, and at most their execs cannor travel to China.

This suggests my "no-quarter" position may be a strong plurality or even a majority.

Another data point: when Pornhub attempted to comy wirh Louisiana's age verificarion law, rhey lost 80% of traffic from that state. Thus they block 24 other age verification states and do not block Tor at this writing.

**Comment by ronanhigg** [comment]
@lukefromdc and @trwnh thank you for your interesting comments.   I respect your viewpoints.  

The majority of Mastodon server owners / admins are going to want to ensure that they are in compliance with the various age verification laws around the world.   These new laws have been enacted by democratically elected officials with a mandate from their constituents.     Mastodon server owners who reside in jurisdictions that now require age verification (EU, UK, Australia) need to be supported.

It is only fair that server admins who don't want to break the law are empowered with the tools to enable them to comply with the law.  And it is important that the tools they are given by the Mastodon developer community adhere to the Fediverse ethos of privacy and security.   That is why Veritage is offering our zero knowledge proof system here.

Given a choice between age verification solutions that require users to upload and transmit their identity documents versus solutions that don't require that, it behooves Mastodon support the latter.   

If some server managers choose to not enable age verification on their server at all, incliuding in jurisdictions that legally require it, that is the server manager's choice.   But the Mastodon server owners deserve a choice.

**Comment by colinstu12** [comment]
> majority of Mastodon server owners / admins are going to want to ensure that they are in compliance

I wouldn't be so sure of that.

**Comment by lukefromdc** [comment]
I think you can forget compliance from servers run by political activist groups that are ideologically
opposed to age verification. Note that if it ever becomes mandatory in the server side code,
admins opposing it can just not update, or recompile the code with the offending commit reverted,

**Comment by PhilGPk** [comment]
> [...] Mastodon server owners who reside in jurisdictions that now require age verification (EU, UK, Australia) need to be supported. [...] If some server managers choose to not enable age verification on their server at all, incliuding in jurisdictions that legally require it, that is the server manager's choice. But the Mastodon server owners deserve a choice.

I take the view that even though I don't like the current legislative requirements, being able to comply with them in the most privacy preserving way feasible would be a valuable optional feature for server owners.

It would also make it easier to advocate for some organisations to use Mastodon servers for their social media instead of the likes of Facebook et al.  They won't have a choice about compliance.

**Comment by ronanhigg** [comment]
> I think you can forget compliance from servers run by political activist groups that are ideologically
> opposed to age verification. Note that if it ever becomes mandatory in the server side code,
> admins opposing it can just not update, or recompile the code with the offending commit reverted,

Yes @lukefromdc , that's why I believe it is important that age assurance is made available to Mastodon server admins as an **_option_** in the Admin / Settings / Registrations.   Similar to how hCAPTCHA was implemented as an optional setting with a checkbox to enable or disable it.   (screenshot attached)

<img width="855" height="600" alt="Image" src="https://github.com/user-attachments/assets/f1659087-f035-41fb-8ded-d71c815bf201" />

**Comment by lukefromdc** [comment]
As long as it stays optional and every nation doesn't have their own Great Firewall of whoever with
China-level resources this should not be a big problem. A corporate server used instead of Facebook
to avoid leaking PII to a rival company has very different needs than say, an antifa server for frontline
activists fighting ICE in Minneapolis. That the same code can serve both is actually rather impressive.

I don't like the precedent, but there is also the fact that if servers like mastodon.social shut down, their
users might go back to Instagram instead of to less well known servers like the ones I am familiar with.

My main worry is avoiding the need for server admins on my side of the house to patch and rebuild
every update to remove code that does NOT provide the option to say no. Some folks who know how
to run a server but are not familiar with pulling from a git repo, reverting commits (sometimes manually when
automation fails), setting up the build environment, and finally building the code might deem this to be
more than they can handle. 

A possible future problem:
What do you think the odds are of servers ending up only able to federate with servers with the same
age verification policy as themselves a few years down the road? This would split the Fediverse in half-
and users may be able to only see half of whichever type of server they are on. 

I suspect the best brake against this may well be  the fact that the current proposals do NOT support
having an underage-only, censored version for under-18 (or whatever) users, so there is little real value 
for governments in blocking posts from non-age verifying servers from being read from an age verifying server.  
Really extreme regimes might worry about non age verifying servers accessing material posted under age 
verification though. If the noncomplying server are outside their jurisdictions they might not spend energy 
on it though, just as the US FOSTA/SESTA laws concerning sex workers have been effectively defeated by 
migration of sex worker sites out of the US.

I suppose the people behind FOSTA/SESTA would blow their top if craigslist.com was able to federate with
2backpage.com, thus getting back sex worker classifies without hosting them in the US. Difference is any
content that can be posted in defiance of age verification only can also be posted by an age verified
user on a complying server. This part of the question would come up if Chat Control targets EU Mastodon
servers though. Chat Control is so intrusive EU servers might have to move or shut down, and privacy/
security minded servers could not safely federate with any chat control compliant server due to the risk
of content frpm their servers getting scanned.

**Comment by Sunconure11** [comment]
https://www.youtube.com/watch?v=ylAP4Oigsow

**Comment by ProGamerGov** [comment]
Any sort of "verification" for age represents an unacceptable privacy violation. There is no such things a privacy protecting or anonymous age verification/age assurance.

If the user selecting their age when making an account isn't good enough, then those countries are free to try and block Mastodon.

**Comment by ronanhigg** [comment]
@ProGamerGov could you elaborate on why you think there is no such thing as a privacy protecting or anonymous age verification?   [In my comment above I discussed Zero Knowledge Proof cryptography](https://github.com/mastodon/mastodon/issues/24440#issuecomment-3932347491) as just such a thing.   Where do you see the lack of privacy protection or lack of anonymity in that technology?

The issue here is not whether countries will try and block a specific Mastodon server, but whether or not server owners should be empowered with the option to comply with the law in their jurisdiction.   Countries won't bother trying to block the Fediverse, they will simply issue fines to businesses (or citizens) who are breaking the law.   

I live in a country that will probably issue my business a fine if I do not comply with the incoming age verification laws.   I want the option to be able to enable AV on my servers.

> Any sort of "verification" for age represents an unacceptable privacy violation. There is no such things a privacy protecting or anonymous age verification/age assurance.
> 
> If the user selecting their age when making an account isn't good enough, then those countries are free to try and block Mastodon.

**Comment by lukefromdc** [comment]
What matters to me is the option to exclusively use servers that DO NOT comply with
age verification. They can be in an unaffected jurisdiction and run by someone who
never travels internationally, they can be managed over Tor in defiance of these
unjust laws, whatever.

I refuse to use any online resource that demands ID. I would sooner go 100% offline
but thankfully there is the dark web.

**Comment by ronanhigg** [comment]
@lukefromdc you seem to be missing the key point that Zero Knowledge Proof technology does not require ID.   A website (or server) that needs to know that a user is above a certain age does not need to see an ID.   They just need a Zero Knowledge Proof.    Don't take my word for it, here's some AI mode blurb https://share.google/aimode/S5HEfy2rw4ZFcLAxn

Just because the very large online platforms like Facebook, TikTok and even Discord have opted to partner with age verification companies (Persona, Yoti, k-ID) that do demand ID, does not mean all websites have to.  There are alternatives that are privacy and anonymity protecting.

**Comment by lukefromdc** [comment]
Again-for my own purposes I simply do not care. Same as refusing to buy age-restricted items in
physical stores. I oppose age verification by ANY means, so I personally will not cooperate with it
in any way. I will shut down any account that asks for any form of proof of age, including my Github
account if it comes to that. No exceptions, Zero Voluntary Cooperation.

**Comment by ProGamerGov** [comment]
@ronanhigg The problem with ZKP is that it was never designed for this task. It requires the use of a third party to perform the verification, and the problem of collusion (like an encryption backdoor) is impossible to prevent. ZKP systems are only anonymous in theory, if you assume the third party is never going to act maliciously. In practice, such a system makes tracking extremely easy and its pretty much guaranteed to be abused for surveillance and spying.

Another issue is that the implementations almost always have anti-tampering requirements that force users to install invasive and non-open source software. As an example, the EU's Digital Identity Wallet requires users install GooglePlay Services / IOS equivalent, and bans users from rooting/jailbreaking their devices. There are other privacy problems with the EU's strategy, like only giving users a small batch of single tokens that expire after 3 months and requiring more verification to obtain more tokens. but those are some of the main ones. And none of this even touches on what the age verification company does with your personal information.

**Comment by trwnh** [comment]
> none of this even touches on what the age verification company does with your personal information.

or more importantly, what the consequences of age verification are to society and freedom. the serious harms it poses should be given far more weight instead of trying to comply with unjust laws in advance.

**Comment by ronanhigg** [comment]
@ProGamerGov I think you might be mistaken in your comment above.  Zero Knowledge Proof does not require a third party for the verification.  It _can_ use a third party for verification, but it's not required.   For example if you look at the ZKpassport project, the Relying Party (website looking for proof of age) adds code to their server that enables them to receive and verify a proof sent from the prover (user).   https://zkpassport.id

You might be conflating what the EU Digital Identity wallet has incorrectly been calling zero knowledge proof, selective disclosure Verifiable Credentials (SD-JWT), with a true Zero Knowledge Proof.   In the EUDIW project, they have (at least until recently) been using  the"zero knowledge" terminology in relation to Issuer signed claims of credentials such as ageOver18=true.   In some ways, this is a way to prove you are over 18 without sending any further information, but it's not _mathematically_ a Zero Knowledge Proof.   

To make the issue more confusing, the EUDIW project is now beginning to develop an actual Zero Knowledge Proof implementation using the Longfellow ZK library that Google open sourced last year.  https://github.com/eu-digital-identity-wallet/av-lib-ios-longfellow-zkp

This kind of terminology abuse happens when you let bureaucrats lead the charge. LOL.

But even with the EUDIW project, the Verifier does not need to be a third party.   A website can deploy the Verifier server code for their own, and only their own, use to verify Verifiable Credentials sent from a wallet.


I think in any discussion of age verification using digital identity wallets, it is important to be clear on the technical design differences between Verifiable Credentials infrastructure and Zero Knowledge Proof; and between Self-Sovereign Identity wallets and Federated Identity wallets.   


In relation to the requirement for user to install "invasive and non-open source software".   That is not always the case.   Again the EUDIW project is a good example, where the app that is installed is completely open source.   The wallet app does have to leverage some of the secure enclave cryptographic services on the smartphone device to ensure the software has not been tampered with, but this does not mean the use of the software can be tracked by the Issuer of the Verifiable Credentials.  It should be clear why a wallet app has to ensure that it and the OS it runs on are closed systems: to ensure that fake credentials cannot be installed.

That being said, in some commercial and proprietary implementations of Verifiable Credentials, the wallet software is not open source: for example: Google Wallet, Apple Wallet.  


On your point "other privacy problems with the EU's strategy, like only giving users a small batch of single tokens that expire after 3 months and requiring more verification to obtain more tokens"   I agree.   Not only is this a pain for consumers, but it also introduces an attack vector (even though it is being done to try and increase security and privacy).   

In our Veritage Zero Knowledge Proof implementation, we don't expire the underlying credential from which the proofs are generated because your date of birth never changes.


In relation to the portion of Mastodon server owners who would like the ability to comply with age verification laws, the idea here is that if they enable age verification on their server administration console, then the server will run as a Verifier -- no third party, man-in-the-middle required.   More than that, even the server owner does not get access to sensitive personal identity information other than the proof that the user is over16=true.   The user is protected from the server owner and other prying eyes.  The server owner is protected from the law.

**Comment by scytaleqc** [comment]
@ronanhigg You might be missing that @trwnh 's point is the **first** you should address:

> what the consequences of age verification are to society and freedom. the serious harms it poses should be given far more weight instead of trying to comply with unjust laws in advance.

For a lot of people, no technology will matter if the law's just plain nefarious.

**Comment by ronanhigg** [comment]
@scytaleqc the GitHub issue is in relation to the OP suggestion: "There should be some option on account creation that lets instance owners redirect users to a third-party age-verification provider and accept their tokens in order to proceed. "   

There are other forums to discuss the consequences of laws on society, with the seriousness and weight the topic deserves.   

It doesn't matter that I think uploading sensitive identity documents to third party verifiers like Persona or k-ID/Veratad is a bad idea; and that I am being self-serving in promoting our company's alternatives.

It doesn't matter here that I think there are many areas where society benefits from age checks: applying for a driver license, getting married, purchasing a gun, signing a contract, registering to vote, running for political office, starting to work, drinking alcohol, smoking tobacco.   Even if these age checks limit freedom.

What deos matter is ensuring that Mastodon server owners are given the option to comply with the laws in their jurisdiction; and that the tools they are given are aligned with the ethos of the Fediverse.

It will be a Discord-level tragedy if Mastodon server developers deploy age verification options that are privacy invasive.

**Comment by lukefromdc** [comment]
It is also important that users have the option (as they do now) to self-host without age verification
or use a noncomplying/unaffected server.

**Comment by trwnh** [comment]
> It will be a Discord-level tragedy if Mastodon server developers deploy age verification options that are privacy invasive.

It will be a tragedy if they do it at all, period.

There's a reason this GitHub issue currently has only 1 thumbs up and 24 thumbs down, and it's not strictly about how "privacy-preserving" is the mechanism being used. The "ethos of the Fediverse" is already countercultural and also will not take kindly to building out the social infrastructure needed to cut off minors from their support networks and limit access to information deemed too "age-inappropriate", like the fact that queer people exist. The threat isn't that your Mastodon admin will use your age signal to serve you ads or sell your information; the threat is that the very concept of "age verification" becomes societally acceptable and normalized, which would be disastrous to anonymity and information freedom, and all that the loss of these things entails.

**Comment by lukefromdc** [comment]
I personally will be calling for boycotts and migration away from complying servers. This is specifically
to protect LGBTQ youth by ensuring that the bulk of the Fediverse remains accessable. Those who
are not up for running a noncomplying server in an age verification jurisdiction won't be breaking any
laws if they first post a guide to migration to other servers, then shut down their own before this can
take effect.

**Comment by dsntechadmin** [comment]
> ### Pitch
> 
> Mastodon may need a way to verify the ages of users where required by law. There should be some option on account creation that lets instance owners redirect users to a third-party age-verification provider and accept their tokens in order to proceed. Having such a facility would also allow a method of screening NSFW content marked as such (see [#9468](https://github.com/mastodon/mastodon/issues/9468) ).
> 
> Additionally, instance owners should be able to restrict access from certain regions if they are unable to implement such controls (due to the associated costs).
> 
> ### Motivation
> 
> Several countries and US states have enacted age verification requirements for social media sites. This means teens would not be permitted to access social media sites without either proving they are over specified age or providing parental consent.



> ### Pitch
> 
> Mastodon may need a way to verify the ages of users where required by law. There should be some option on account creation that lets instance owners redirect users to a third-party age-verification provider and accept their tokens in order to proceed. Having such a facility would also allow a method of screening NSFW content marked as such (see [#9468](https://github.com/mastodon/mastodon/issues/9468) ).
> 
> Additionally, instance owners should be able to restrict access from certain regions if they are unable to implement such controls (due to the associated costs).
> 
> ### Motivation
> 
> Several countries and US states have enacted age verification requirements for social media sites. This means teens would not be permitted to access social media sites without either proving they are over specified age or providing parental consent.

**Comment by dsntechadmin** [comment]
> OK here is what I have been doing behind the scenes in Australia.

I have been reading this thread while I inputting my opinion, now I have read some serious discussions and some degrading comments.. this thread was meant to be a discussion and not an all out sledge fest.

BUT now I have founded the Online Broadcasting Industry Association(O.B.I.A.) and that means Australian streamers, news media and SOCIAL media (aka mastodon and Bluesky plus small time owners)

This means we have the body to try and stop this from getting worse and maybe reverse this ban!

I think this esafety commissioner is still a meta shill aka the official meta and X spy.  And she was actually unelected as by our constitution states the position must be elected by 2/3 of parliament.  But there is also a major problem.  Federated or defederated, mastodon may be hit worldwide not just Australian servers.  This could be catastrophic for all owners of mastodon and federated socials. No one wants a $49million fine at all!

So if you live in Australia, I made a e-Petition stopping this from happening and the link is here:

https://www.aph.gov.au/e-petitions/petition/EN9614

If this petition does not get 500 signatures we will fail.  So please spread this link to EVERYONE on all socials.. YouTube PeerTube X Facebook heck! even Snapchat!

There is also one more route but it means to pay $100000 in legal fees as we need to take this to the high court.  

EVERYONE here make an e-petition in your local area to reverse this BS..  

PS please try to keep this thread civil as I don’t know.. Brodie Robertson could see this thread.. or SomeOrdinaryGamers or even Penguinz0 as I am trying to share this with them…

<img width="585" height="1266" alt="Image" src="https://github.com/user-attachments/assets/59ef7f0c-5fe2-4d28-af62-8af029328f61" />

**Comment by shaedrich** [comment]
> In relation to the requirement for user to install "invasive and non-open source software". That is not always the case. Again the EUDIW project is a good example, where the app that is installed is completely open source. The wallet app does have to leverage some of the secure enclave cryptographic services on the smartphone device to ensure the software has not been tampered with, but this does not mean the use of the software can be tracked by the Issuer of the Verifiable Credentials. It should be clear why a wallet app has to ensure that it and the OS it runs on are closed systems: to ensure that fake credentials cannot be installed.

@ronanhigg Yeah, well, having to have GooglePlayServices installed ALWAYS at least leaves a risk of being tracked. And most circulating smartphones are Android phones or iPhones, including refurbished ones that might use an alternative mobile OS.

> There should be some option on account creation that lets instance owners redirect users to a third-party age-verification provider

So if there would be absolutely no way around implementing it due to legislation, Mastodon should not allow servers to add just any age verification provider, they like.

> […] uploading sensitive identity documents to third party verifiers like Persona or k-ID/Veratad is a bad idea; […] It will be a Discord-level tragedy if Mastodon server developers deploy age verification options that are privacy invasive.

Just like with Supply Chain Due Diligence Act, Mastodon at least should only allow providers with the same values as Mastodon or do the age verification themselves.

> Zero Knowledge Proof does not require a third party for the verification. It can use a third party for verification, but it's not required.

**Comment by lukefromdc** [comment]
Well, I am one of those who refuse to allow Google Play Services on any phone I use. The important thing for me at this point is that noncomplying Mastodon instances do not have to either fork or stop updating the Mastodon code to block a compulsory age verification requirement for every instance. I can live with splitting the Fediverse if necessary, where age verifying servers talk only to other age verifying servers and noncomplying/non affected jurisdiction servers talk only to oneanother.

Another reason for this is that use of the code for self-hosted single-user instances does not violate any of the current age verification laws to my knowledge, as the user is not posting to, creating an account with, or directly accessing any multi-user server.  Such single-users not the Mastdon software are probably the "service providers" in question.  If this applies to federation of single-user instanes, than self-hosted users could get a checkbox to select which portion of the Fediverse they wish to connect to.

As I have said before, I am publicly refusing to verify my age or identity to use any website, so I won't be staying on any compliant server. BTW, at this point this Github account and possibly my Gitlab and GNOME gitlab accounts are my only remaining accounts inside the US. If they go age verification I will be signing off of course.

Personally, I would prefer to see servers in age verification jurisdictions either move, go underground (if operated by folks with no assets reachable by courts), or shut down.
